source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,275 @@ extern "C"
+      // the above functions (see metish.h)
+      AssertThrow (ierr == 1, ExcMETISError (ierr));
+
+      // now copy back generated indices into the output array
+      std::copy (int_partition_indices.begin(),
+                 int_partition_indices.end(),
+                 partition_indices.begin());
+#endif
+    }
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    //Query functions for partition_zoltan
+    int get_number_of_objects(void *data, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+
+      *ierr = ZOLTAN_OK;
+
+      return graph->n_rows();
+    }
+#endif
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    void get_object_list(void *data, int sizeGID, int sizeLID,
+                         ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                         int wgt_dim, float *obj_wgts, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      Assert( globalID != nullptr , ExcInternalError() );
+      Assert( localID != nullptr , ExcInternalError() );
+
+      //set global degrees of freedom
+      auto n_dofs = graph->n_rows();
+      for (int i=0; i < n_dofs; i++)
+        {
+          globalID[i] = i;    //global ids run from 0 to n_dofs-1
+          localID[i] = i;     //same as global ids
+        }
+
+      return;
+    }
+#endif
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    void get_num_edges_list(void *data, int sizeGID, int sizeLID,
+                            int num_obj,
+                            ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                            int *numEdges, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+
+      *ierr = ZOLTAN_OK;
+
+
+      Assert ( numEdges != nullptr , ExcInternalError() );
+
+      for (int i=0; i<num_obj; ++i)
+        numEdges[i] = graph->row_length(globalID[i])-1;
+      return;
+    }
+#endif
+
+//Query function unused if zoltan is not installed
+#ifdef DEAL_II_TRILINOS_WITH_ZOLTAN
+    void get_edge_list(void *data, int sizeGID, int sizeLID,
+                       int num_obj, ZOLTAN_ID_PTR globalID, ZOLTAN_ID_PTR localID,
+                       int *num_edges,
+                       ZOLTAN_ID_PTR nborGID, int *nborProc,
+                       int wgt_dim, float *ewgts, int *ierr)
+    {
+      SparsityPattern *graph = reinterpret_cast<SparsityPattern *>(data);
+      *ierr = ZOLTAN_OK;
+
+      ZOLTAN_ID_PTR nextNborGID = nborGID;
+      int *nextNborProc = nborProc;
+
+      Assert( nextNborGID != nullptr , ExcInternalError() );
+      Assert( nextNborProc != nullptr , ExcInternalError() );
+
+      //Loop through rows corresponding to indices in globalID implicitly
+      for (int i=0; i < num_obj; ++i)
```

Comment:
`for (SparsityPattern::size_type i=0; i <  static_castSparsityPattern::size_type>(num_obj); ++i)` to avoid

```
/home/bob/source/source/lac/sparsity_tools.cc:219:20: warning: comparison of integers of different signs: 'int' and 'size_type' (aka 'unsigned int') [-Wsign-compare]
            if ( i != col->column() )
                 ~ ^  ~~~~~~~~~~~~~
```